qiskit/circuit/operation.py

```diff
@@ -0,0 +1,62 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Quantum Operation Mixin."""
+
+from abc import ABC
+from qiskit.circuit.exceptions import CircuitError
+
+
+class Operation(ABC):
+    """Quantum Operation Mixin Class.
+    For objects that can be added to a :class:`~qiskit.circuit.QuantumCircuit`.
+    These objects include :class:`~qiskit.circuit.Gate`, :class:`~qiskit.circuit.Reset`,
+    :class:`~qiskit.circuit.Barrier`, :class:`~qiskit.circuit.Measure`,
+    and operators such as :class:`~qiskit.quantum_info.Clifford`.
+    """
+
+    # pylint: disable=unused-argument
+    def __new__(cls, *args, **kwargs):
```

Comment:
In qiskit-experiments we found that adding a `__new__` method breaks the class documentation and IDE autocomplete for all classes that subclass because it overrides the init documentation in the class doc string. I would be hesitant about doing this for something like this safety check which seems unnecessary.

Reply:
I'm not sure that `__new__` and `__init__` make this not a mixin. My understanding of a mixin is that it may have a mix of concrete and abstract methods, can't be instantiated, and has no internal state. [ref.](https://rednafi.github.io/digressions/python/2020/07/03/python-mixins.html#differences-between-interfaces-abstract-classes-and-mixins) Since this has `_name`, etc. it may be behaviing more like an abastract class. However, even if we were ok with that the issue you bring up with the documentation may not be worth the re-implementation hassle in derived classes.